_posts/2017-12-08-totw-123.md

```diff
@@ -0,0 +1,132 @@
+---
+title: "Tip of the Week #123: `absl::optional` and `std::unique_ptr`"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/123
+type: markdown
+---
+
+Originally posted as totw/123 on 2016-09-06
+
+By Alexey Skolov and Etienne Dechamps
+
+## How to store values
+
+This tip discusses several ways of storing values. Here we use class member
+variables as an example, but many of the points below also apply to local
+variables.
+
+```c++
+#include <memory>
+#include "third_party/absl/types/optional.h"
+#include ".../bar.h"
+
+class Foo {
+  ...
+ private:
+  Bar val_;
+  absl::optional<Bar> opt_;
+  std::unique_ptr<Bar> ptr_;
+};
+```
+
+### As a bare object
+
+This is the simplest way. `val_` is constructed and destroyed at the beginning
+of `Foo`'s constructor and at the end of `Foo`'s destructor, respectively. If
+`Bar` has a default constructor, it doesn't even need to be initialized
+explicitly.
+
+`val_` is very safe to use, because its value can't be null. This removes a
+class of potential bugs.
+
+But bare objects are not very flexible:
+
+*   The lifetime of `val_` is fundamentally tied to the lifetime of its parent
+    `Foo` object, which is sometimes not desirable. If `Bar` supports move or
+    swap operations, the contents of `val_` can be replaced using these
+    operations, while any existing pointers or references to `val_` continue
+    pointing or referring to the same `val_` object (as a container), not to the
+    value stored in it.
+*   Any arguments that need to be passed to `Bar`'s constructor need to be
+    computed inside the initializer list of `Foo`'s constructor, which can be
+    difficult if complicated expressions are involved.
+
+### As `absl::optional<Bar>`
+
+This is a good middle ground between simplicity of bare objects and flexibility
```

Comment:
Nit: simplicity -> the simplicity, flexibility -> the flexibility

Reply:
Done.